### PR TITLE
feat(key-auth) optionally search the request body for credentials

### PR DESCRIPTION
**Pull request by p0pr0ck5**
### Summary

This commit provides an optional plugin configuration field to
search the request body for the authentication credentials. This
behavior respects the hide_credentials field.

This commit also expands the access test suite to cover a few more
use cases.

### Full changelog

* add the key_in_body configuration option
* add related tests
* expand config.hide_credentials tests to cover all use cases (headers, query string, req body)

### CODE REVIEW

File: spec/03-plugins/10-key-auth/02-access_spec.lua [hunk 1]
```diff
@@ -183,32 +271,83 @@ describe("Plugin: key-auth (access)", function()
...
+          local res = assert(client:send(harness[type][1]))
+          local body = assert.res_status(200, res)
+          local json = cjson.decode(body)
+
+          -- small workaround for how mockbin sends body data
+          local field
+          if type == "postData" then
+            local t = json[type].text:sub(8)
+            field = { apikey = t ~= "" and t or nil }
+          else
```
Comment (thibaultcha) [inline spec/03-plugins/10-key-auth/02-access_spec.lua]: I think we should apply our codestyle to tests as well

File: spec/03-plugins/10-key-auth/02-access_spec.lua [hunk 2]
```diff
@@ -183,32 +271,83 @@ describe("Plugin: key-auth (access)", function()
...
+        it("true doesn't send key to upstream", function()
+          local res = assert(client:send(harness[type][2]))
+          local body = assert.res_status(200, res)
+          local json = cjson.decode(body)
+
+          local field
+          if type == "postData" then
+            local t = json[type].text:sub(8)
+            field = { apikey = t ~= "" and t or nil }
+          else
```
Comment (thibaultcha) [inline spec/03-plugins/10-key-auth/02-access_spec.lua]: ditto

**Comment by p0pr0ck5** [comment]
Thanks, minor style fixes applied and merged.

**Comment by thibaultcha** [comment]
I made a slight oversight: we should have updated the `"No API key found in ..."` error message returned on HTTP 401, since we now take the body in consideration as well. Maybe even better, we could change it to simply: `"No API key provided"` or something similar. After all, where else than the headers, the querystring, and the body could once insert an arbitrary value? Not that many places...

**Comment by p0pr0ck5** [comment]
@thibaultcha would the resulting change in the return values constitute a breaking change, and thus belong in the next major release? or this is something we're comfortable patching now?

**Comment by thibaultcha** [comment]
I think as long as the HTTP status code doesn't change, this is hardly considered a breaking change. It should be fine.

**Comment by p0pr0ck5** [comment]
:+1: wasn't sure about our stance on the message content itself (i imagine a UI application working based on the contents of the `message` field). done in #2502.

**Pull request merged**